File: src/main/java/org/reactome/server/tools/DuplicateChecker.java
package org.reactome.server.tools;

import com.martiansoftware.jsap.JSAPResult;
import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVPrinter;
import org.reactome.server.tools.model.Icon;
import org.reactome.server.tools.model.Reference;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.xml.bind.JAXBContext;
import javax.xml.bind.JAXBException;
import javax.xml.bind.Unmarshaller;
import java.io.BufferedWriter;
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;


@SuppressWarnings("unused")
public class DuplicateChecker implements Checker {
    private static final Logger errorLogger = LoggerFactory.getLogger("errorLogger");
    private final JSAPResult config;

    public DuplicateChecker(JSAPResult config) {
        this.config = config;
    }

    private int errorDuplicate = 0;
    private int xmlNum = 0;

    @Override
    public void process() {
        String directory = config.getString("directory");

        File filesInDir = new File(directory);

        File[] xmlFiles = filesInDir.listFiles((dir, name) -> name.endsWith(".xml"));

        List<String> duplicatedReferenceId = duplicatedId(xmlFiles);

        try {
            // write potential duplicate item into a csv file
            File csvFile = new File("potential_duplicate.csv");
            BufferedWriter writer = Files.newBufferedWriter(Paths.get(csvFile.getPath()));
            CSVPrinter csvPrinter = new CSVPrinter(writer, CSVFormat.DEFAULT.withHeader("Name", "Category", "Reference", "File"));

            if (xmlFiles != null) {
                for (File file : xmlFiles) {
                    xmlNum = xmlFiles.length;
                    Icon icon = convertXmlToObj(file);
                    if (icon != null) {
                        findDuplicated(file, icon, duplicatedReferenceId, csvPrinter);
                    }
                }
            }
            csvPrinter.flush();
            csvPrinter.close();
        } catch (IOException e) {
            errorLogger.error("Could not write potential_duplicate.csv");
            e.printStackTrace();
        }

    }

    private Icon convertXmlToObj(File file) {
        JAXBContext jaxbContext;
        try {
            jaxbContext = JAXBContext.newInstance(Icon.class);
            Unmarshaller jaxbUnmarshaller = jaxbContext.createUnmarshaller();
            return (Icon) jaxbUnmarshaller.unmarshal(file);
        } catch (JAXBException ignored) {
        }
        return null;
    }

    private void findDuplicated(File xmlFile, Icon icon, List<String> duplicated, CSVPrinter csvPrinter) throws IOException {

        List<Reference> references = icon.getReferences();

        if (references != null) {
            for (Reference reference : references) {
                if (duplicated.contains(reference.getId())) {
                    csvPrinter.printRecord(icon.getName(), icon.getCategories().toString().replace("[", "").replace("]", ""), reference.getId(), xmlFile.getName());
                }
            }
        }
    }

    private List<String> duplicatedId(File[] xmlFiles) {
        List<String> referenceId = new ArrayList<>();
        if (xmlFiles != null) {
            for (File file : xmlFiles) {
                Icon icon = convertXmlToObj(file);
                if (icon != null && icon.getReferences() != null && icon.getReferences().size() > 0) {
                    for (Reference reference : icon.getReferences()) {
                        referenceId.add(reference.getId());
                    }
                }
            }
        }
        //get duplicate items

        return referenceId.stream()
                .filter(e -> Collections.frequency(referenceId, e) > 1)
                .distinct()
                .collect(Collectors.toList());
    }

    @Override
    public int getFailedChecks() {
        return errorDuplicate;
    }

    @Override
    public int getTotalChecks() {
        return xmlNum;
    }
}

File: src/main/java/org/reactome/server/tools/DuplicateChecker.kt
package org.reactome.server.tools

import com.martiansoftware.jsap.JSAPResult
import org.apache.commons.csv.CSVFormat
import org.apache.commons.csv.CSVPrinter
import org.reactome.server.tools.model.Icon
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths
import java.util.*
import java.util.stream.Collectors
import javax.xml.bind.JAXBContext
import javax.xml.bind.JAXBException

@Suppress("unused")
class DuplicateChecker(private val config: JSAPResult) : Checker {
    var _failedChecks: Int = 0
    var _totalChecks: Int = 0

    override fun getFailedChecks(): Int {
        return _failedChecks
    }

    override fun getTotalChecks(): Int {
        return _totalChecks
    }

    override fun process() {
        val directory = config.getString("directory")

        val filesInDir = File(directory)

        val xmlFiles = filesInDir.listFiles { _: File?, name: String -> name.endsWith(".xml") }

        val duplicatedReferenceId = xmlFiles?.let { duplicatedId(it) } ?: emptyList()

        try {
            // write potential duplicate item into a csv file
            val csvFile = File("potential_duplicate.csv")
            val writer = Files.newBufferedWriter(Paths.get(csvFile.path))
            val csvPrinter = CSVPrinter(writer, CSVFormat.DEFAULT.withHeader("Name", "Category", "Reference", "File"))

            if (xmlFiles != null) {
                for (file in xmlFiles) {
                    _totalChecks = xmlFiles.size
                    val icon = convertXmlToObj(file)
                    if (icon != null) {
                        findDuplicated(file, icon, duplicatedReferenceId, csvPrinter)
                    }
                }
            }
            csvPrinter.flush()
            csvPrinter.close()
        } catch (e: IOException) {
            errorLogger.error("Could not write potential_duplicate.csv")
            e.printStackTrace()
        }
    }

    private fun convertXmlToObj(file: File): Icon? {
        val jaxbContext: JAXBContext
        try {
            jaxbContext = JAXBContext.newInstance(Icon::class.java)
            val jaxbUnmarshaller = jaxbContext.createUnmarshaller()
            return jaxbUnmarshaller.unmarshal(file) as Icon
        } catch (ignored: JAXBException) {
        }
        return null
    }

    @Throws(IOException::class)
    private fun findDuplicated(xmlFile: File, icon: Icon, duplicated: List<String?>, csvPrinter: CSVPrinter) {
        val references = icon.references

        if (references != null) {
            for (reference in references) {
                if (duplicated.contains(reference.id)) {
                    csvPrinter.printRecord(
                        icon.name,
                        icon.categories.toString().replace("[", "").replace("]", ""),
                        reference.id,
                        xmlFile.name
                    )
                }
            }
        }
    }

    private fun duplicatedId(xmlFiles: Array<File>): List<String> {
        return xmlFiles
            .flatMap { convertXmlToObj(it)?.references ?: emptyList() }
            .map { it.id }
            .groupingBy { it }
            .eachCount()
            .filter { it.value > 1 }
            .keys
            .toList()
    }

    companion object {
        private val errorLogger: Logger = LoggerFactory.getLogger("errorLogger")
    }
}